### ISSUE TITLE
Many-to-Many relation seems to break

### ISSUE THREAD

**Issue by dev-rke** [waterlinejs/sqlite3-adapter#1]
Hi,

i like your new sqlite adapter, it works very well with node 4.x.x.
Unfortunately i stumbled upon an error when using multiple relations:

```
Error (E_UNKNOWN) :: Encountered an unexpected error
: SELECT "requested_services"."id", "requested_services"."label" FROM "requested_services" AS "requested_services"  WHERE "requested_services"."id" IN )  - SQLITE_ERROR: near ")": syntax error
  at Error (native)


Details:  Error: SELECT "requested_services"."id", "requested_services"."label" FROM "requested_services" AS "requested_services"  WHERE "requested_services"."id" IN )  - SQLITE_ERROR: near ")": syntax error
```

This happens after i added two relations at once and then try to save them:

```
model.existing_services.add(['12', '23', '15']);
model.requested_services.add(['21', '22', '23']);

model.save(); // here the error occurs
```

If i only add one single relation, it works very well.
If i swap the two lines which add relations, i get the same error, but with the "existing_services" table.

I defined both relations as many-to-many relations, also using the "dominant" property, but i think it should not be required. I also tried dropping all database tables, but it does not work.

So i assume there might be a bug.
Do you have any ideas?

**Comment by tjwebb** [comment]
Hmm, all the waterline integration tests were passing. I'll check it out when I get a chance. If you have any other ideas in the meantime, let me know.

**Comment by dev-rke** [comment]
Hi tjwebb,

thanks for your time.
I just build a small example and it seems to work out of the box (this is coffeescript, i put a javascript alternative below):

``` coffeescript
Waterline = require "waterline"

class a extends Waterline.Collection
    identity: "a"
    connection: "sqlitedb"

    autoPK: false
    autoCreatedAt: false
    autoUpdatedAt: false

    attributes:
        id:
            type: "integer"
            primaryKey: true
        name:
            type: "string"
        b:
            collection: "b"
            via: "a"
        c:
            collection: "c"
            via: "a"

class b extends Waterline.Collection
    identity: "b"
    connection: "sqlitedb"

    autoPK: false
    autoCreatedAt: false
    autoUpdatedAt: false

    attributes:
        id:
            type: "integer"
            primaryKey: true
        name:
            type: "string"
        a:
            collection: "a"
            via: "b"


class c extends Waterline.Collection
    identity: "c"
    connection: "sqlitedb"

    autoPK: false
    autoCreatedAt: false
    autoUpdatedAt: false

    attributes:
        id:
            type: "integer"
            primaryKey: true
        name:
            type: "string"
        a:
            collection: "a"
            via: "c"

config =
    adapters:
        sqlitedb: require "waterline-sqlite3"
    connections:
        sqlitedb:
            adapter: "sqlitedb"
            type: "disk"
            filename: "./databasetest_sqlite.db"
            debug: false


waterline = new Waterline()
waterline.loadCollection(a)
waterline.loadCollection(b)
waterline.loadCollection(c)
waterline.initialize config, (err, models) ->
    return console.error(err) if err

    a = models.collections.a
    b = models.collections.b
    c = models.collections.c

    firstId = Math.floor(Math.random() * 3000) + 1

    b.create({
        id: firstId
        name: "testb"
    }).then (bModel) ->
        console.log bModel

    c.create({
        id: firstId
        name: "testc"
    }).then (cModel) ->
        console.log cModel

    a.create({
        id: firstId
        name: "testa"
    }).then (aModel) ->
        console.log aModel
        aModel.b.add([firstId])
        aModel.c.add([firstId])
        aModel.save (err, aModelWithRelations) ->
            console.log aModelWithRelations
```

Also if you prefer JavaScript here the compiled code:

``` Javascript
var Waterline, a, b, c, config, waterline,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

Waterline = require("waterline");

a = (function(superClass) {
  extend(a, superClass);

  function a() {
    return a.__super__.constructor.apply(this, arguments);
  }

  a.prototype.identity = "a";
  a.prototype.connection = "sqlitedb";
  a.prototype.autoPK = false;
  a.prototype.autoCreatedAt = false;
  a.prototype.autoUpdatedAt = false;

  a.prototype.attributes = {
    id: {
      type: "integer",
      primaryKey: true
    },
    name: {
      type: "string"
    },
    b: {
      collection: "b",
      via: "a"
    },
    c: {
      collection: "c",
      via: "a"
    }
  };

  return a;

})(Waterline.Collection);

b = (function(superClass) {
  extend(b, superClass);

  function b() {
    return b.__super__.constructor.apply(this, arguments);
  }

  b.prototype.identity = "b";
  b.prototype.connection = "sqlitedb";
  b.prototype.autoPK = false;
  b.prototype.autoCreatedAt = false;
  b.prototype.autoUpdatedAt = false;

  b.prototype.attributes = {
    id: {
      type: "integer",
      primaryKey: true
    },
    name: {
      type: "string"
    },
    a: {
      collection: "a",
      via: "b"
    }
  };

  return b;

})(Waterline.Collection);

c = (function(superClass) {
  extend(c, superClass);

  function c() {
    return c.__super__.constructor.apply(this, arguments);
  }

  c.prototype.identity = "c";
  c.prototype.connection = "sqlitedb";
  c.prototype.autoPK = false;
  c.prototype.autoCreatedAt = false;
  c.prototype.autoUpdatedAt = false;

  c.prototype.attributes = {
    id: {
      type: "integer",
      primaryKey: true
    },
    name: {
      type: "string"
    },
    a: {
      collection: "a",
      via: "c"
    }
  };

  return c;

})(Waterline.Collection);

config = {
  adapters: {
    sqlitedb: require("waterline-sqlite3")
  },
  connections: {
    sqlitedb: {
      adapter: "sqlitedb",
      type: "disk",
      filename: "./databasetest_sqlite.db",
      debug: false
    }
  }
};

waterline = new Waterline();
waterline.loadCollection(a);
waterline.loadCollection(b);
waterline.loadCollection(c);

waterline.initialize(config, function(err, models) {
  var firstId;
  if (err) {
    return console.error(err);
  }

  a = models.collections.a;
  b = models.collections.b;
  c = models.collections.c;

  firstId = Math.floor(Math.random() * 3000) + 1;

  b.create({
    id: firstId,
    name: "testb"
  }).then(function(bModel) {
    return console.log(bModel);
  });

  c.create({
    id: firstId,
    name: "testc"
  }).then(function(cModel) {
    return console.log(cModel);
  });

  return a.create({
    id: firstId,
    name: "testa"
  }).then(function(aModel) {
    console.log(aModel);
    aModel.b.add([firstId]);
    aModel.c.add([firstId]);
    return aModel.save(function(err, aModelWithRelations) {
      return console.log(aModelWithRelations);
    });
  });
});
```

So there must be an error i produced or its something weird that only happens in a specific constellation... :-/
Is there any way to find out where the error occurred?
Are there some integrated "debug" options i can turn on?
Is there some workflow description in which order the database queries perform?
I just get an SQL error which does not seem a valid result to me and i dont get any further information where it happened or why it happened.

**Comment by dev-rke** [comment]
Another short question: is there something like transactions, i did not found it in the docs?
It took 16 seconds to import around 200 entries with this simple relations - thats quite long time, even for sqlite.

Edit: ok, i saw i have to implement it myself - but it would be great if that feature would make it into the waterline core to have abstract methods for that.

**Comment by dev-rke** [comment]
TLDR: 
problem in record save() method: optional relations produce errors

Ok, i think i got the source of the problem:
I always implemented the save() method in my original code like this:

``` coffeescript
aModel.save (err, aModelWithRelations) ->
  if err
    console.error err
```

That catches any errors and outputs to console.

I found out that there is an error when i have two relations from my aModel (one to model b and one to model c) and define only one relation (e.g. model b), but ignore setting the other relation (model c).
So i get the error regarding model c.

Full code with description:

``` coffeescript
Waterline = require "waterline"
async = require "async"

class a extends Waterline.Collection
    identity: "a"
    connection: "sqlitedb"

    autoPK: false
    autoCreatedAt: false
    autoUpdatedAt: false

    attributes:
        id:
            type: "integer"
            primaryKey: true
        name:
            type: "string"
        b_model:
            collection: "b_model"
            via: "a"
        c_model:
            collection: "c_model"
            via: "a"

class b_model extends Waterline.Collection
    identity: "b_model"
    connection: "sqlitedb"

    autoPK: false
    autoCreatedAt: false
    autoUpdatedAt: false

    attributes:
        id:
            type: "integer"
            primaryKey: true
        name:
            type: "string"
        a:
            collection: "a"
            via: "b_model"


class c_model extends Waterline.Collection
    identity: "c_model"
    connection: "sqlitedb"

    autoPK: false
    autoCreatedAt: false
    autoUpdatedAt: false

    attributes:
        id:
            type: "integer"
            primaryKey: true
        name:
            type: "string"
        a:
            collection: "a"
            via: "c_model"

config =
    adapters:
        sqlitedb: require "waterline-sqlite3"
    connections:
        sqlitedb:
            adapter: "sqlitedb"
            type: "disk"
            filename: "./databasetest_sqlite.db"
            debug: false

waterline = new Waterline()
waterline.loadCollection(a)
waterline.loadCollection(b_model)
waterline.loadCollection(c_model)
waterline.initialize config, (err, models) ->
    return console.error(err) if err

    a = models.collections.a
    b_model = models.collections.b_model
    c_model = models.collections.c_model

    # flush database
    a.query "DELETE FROM a;"
    a.query "DELETE FROM a_b_model__b_model_a;"
    a.query "DELETE FROM a_c_model__c_model_a;"
    a.query "DELETE FROM b_model;"
    a.query "DELETE FROM c_model;"


    bModels = []
    cModels = []
    aModels = []
    newModels = for x in [1..10]
        bModels.push
            id: x
            name: "testb"
        cModels.push
            id: x
            name: "testc"
        aModels.push
            id: x
            name: "testa"

    async.parallel [
        (callback) ->
            b_model.create(bModels).then (bModel) ->
                callback()

        (callback) ->
            c_model.create(cModels).then (cModel) ->
                callback()

    ], (err, results) ->
        a.create(aModels).then (createdAModels) ->
            for aModel in createdAModels
                aModel.b_model.add([aModel.id])

                # this produces the error, because the save method seems to require both relations set
#               aModel.c_model.add([aModel.id, aModel.id + 1])
                aModel.save (err, aModelWithRelations) ->
                    if err
                        console.error err
```

The data seems to be written to database, even if the error occurs.
So i think this error is not critical, but it should be fixed to avoid confusing users.
I worked two days on this because i did not understand if it is a problem of my code or of the framework.

**Comment by ducky427** [comment]
(Sails and Waterline newbie here)

I've encountered this same issue with a Many-to-Many relationship.

**Comment by tjwebb** [comment]
> Another short question: is there something like transactions,

Waterline doesn't yet support transactions. This is something we're working on.

> So i think this error is not critical, but it should be fixed to avoid confusing users.

Can you pin down which line(s) of code are causing this?

**Comment by ducky427** [comment]
@tjwebb, I am using this through sails and I don't _think_ I am using transactions. But a noob here.

I can create a dummy project with a test which fails with this error if it helps.

cheers!

**Comment by ducky427** [comment]
I've got a project which reproduces this issue. Here it [is](https://github.com/ducky427/waterline-sqlitebug). To reproduce
- npm install -g mocha
- git clone https://github.com/ducky427/waterline-sqlitebug
- cd waterline-sqlitebug
- npm install
- npm test

I get the following error:

```
error: Sending 500 ("Server Error") response:
 Error (E_UNKNOWN) :: Encountered an unexpected error
: SQLITE_ERROR: near ")": syntax error
    at Error (native)
From previous event:
    at Client._query (...../sqlitebug/node_modules/waterline-sqlite3/node_modules/knex/lib/dialects/sqlite3/index.js:80:12)
    at Client.query (...../sqlitebug/node_modules/waterline-sqlite3/node_modules/knex/lib/client.js:127:24)
    at Runner.<anonymous> (...../sqlitebug/node_modules/waterline-sqlite3/node_modules/knex/lib/runner.js:118:24)
From previous event:
    at ...../sqlitebug/node_modules/waterline-sqlite3/node_modules/knex/lib/runner.js:44:21
    at processImmediate [as _immediateCallback] (timers.js:374:17)
From previous event:
    at Runner.run (...../sqlitebug/node_modules/waterline-sqlite3/node_modules/knex/lib/runner.js:30:20)
    at Raw.Target.then (...../sqlitebug/node_modules/waterline-sqlite3/node_modules/knex/lib/interface.js:27:43)
    at Object.<anonymous> (...../sqlitebug/node_modules/waterline-sqlite3/dist/adapter.js:268:20)
    at Object.query (...../sqlitebug/node_modules/waterline-sqlite3/dist/adapter.js:275:8)
    at Object.wrapper [as query] (/...../sqlitebug/node_modules/waterline-sqlite3/node_modules/lodash/index.js:3095:19)
    at ...../sqlitebug/node_modules/waterline-sqlite3/dist/adapter.js:334:21

Details:  Error: SELECT "role"."name", "role"."id", "role"."createdAt", "role"."updatedAt", "role"."application" FROM "role" AS "role"  WHERE "role"."id" IN )  LIMIT 30 - SQLITE_ERROR: near ")": syntax error

      1) create an admin user and GET on /user

  0 passing (1s)
  1 failing

  1) UserController /user create an admin user and GET on /user:
     Error: expected 200 "OK", got 500 "Internal Server Error"
      at Test._assertStatus (node_modules/supertest/lib/test.js:232:12)
      at Test._assertFunction (node_modules/supertest/lib/test.js:247:11)
      at Test.assert (node_modules/supertest/lib/test.js:148:18)
      at Server.assert (node_modules/supertest/lib/test.js:127:12)
      at emitCloseNT (net.js:1518:8)

npm ERR! Test failed.  See above for more details.
```

The query being generated which is causing this issues is:

``` sql
SELECT "role"."name", "role"."id", "role"."createdAt", "role"."updatedAt", "role"."application" FROM "role" AS "role"  WHERE "role"."id" IN )  LIMIT 30
```

If on line 11 of [bootstrap.test.js](https://github.com/ducky427/waterline-sqlitebug/blob/master/test/bootstrap.test.js#L11), I change the connection to `localDiskDb`, the same test passes.

**Comment by ducky427** [comment]
Just FYI, I was able to run the same test using `sails-postgresql` driver and the test passes.

**Comment by ducky427** [comment]
Looks like the call to `find` of `lib/adapter.js` with arguments:

```
options -> { where: { id: [] }, limit: 30 }
```

causes the incorrect sql to be produced

**Comment by dev-rke** [comment]
@tjwebb

> > Another short question: is there something like transactions,
> 
> Waterline doesn't yet support transactions. This is something we're working on.

Yes, i already figured out, thank you.

> > So i think this error is not critical, but it should be fixed to avoid confusing users.
> 
> Can you pin down which line(s) of code are causing this?

Please refer to the last few lines of code of [this](https://github.com/waterlinejs/sqlite3-adapter/issues/1#issuecomment-145383650) comment:

``` coffeescript
                # this produces the error, because the save method seems to require both relations set
#               aModel.c_model.add([aModel.id, aModel.id + 1])
                aModel.save (err, aModelWithRelations) ->
                    if err
                        console.error err
```

I do not understand why the error occurs, in my opinion it should not happen.
